### PR TITLE
feat: implemented use of OpenAI's `tools` API

### PR DESCRIPTION
**Pull request by ErikBjare**
Made an attempt at https://github.com/ErikBjare/gptme/issues/218, was pretty easy to get this minimal thing working.

- [x] get it working for save/append
- [x] get it working for patch
- [x] get it working for ipython
- [x] ~~should we register ipython-functions too?~~ leave it to the ipython tool
- [ ] fix parameter schemas
- [ ] clean up examples prompting
- [ ] get it working for anthropic
- [ ] test it with openrouter/ollama with Llama 1B/3B to see if behavior is improved
   - Addressing #195 

Would improve system prompt adherence, fixing issues like:
 - https://github.com/ErikBjare/gptme/issues/153
 - https://github.com/ErikBjare/gptme/issues/195
 - etc.

----

> [!IMPORTANT]
> Adds support for OpenAI's Tools API, integrating tool usage into chat flow with tool execution and response handling.
> 
>   - **Behavior**:
>     - Adds support for OpenAI's Tools API in `step()` in `chat.py` and `reply()` in `llm.py`.
>     - Parses tool call responses in `step()` using regex to extract tool name and arguments.
>     - Integrates tool execution in chat flow, yielding tool execution results.
>   - **Functions**:
>     - Adds `build_tools_dict()` and `spec2tool()` in `llm.py` to construct tool specifications.
>     - Modifies `reply()`, `_chat_complete()`, and `_stream()` in `llm.py` to accept and handle tools.
>   - **Misc**:
>     - Updates `chat()` and `stream()` in `llm_openai.py` and `llm_anthropic.py` to pass tools to API calls.
>     - Adds `tools` parameter to `chat()` and `stream()` functions in `llm_openai.py` and `llm_anthropic.py`.
>     - Updates `toolcall_re` regex in `tools/base.py` to parse tool calls.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ErikBjare%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 4a11c8b7130750b3d017dc9d0419805dfd61f66b. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: gptme/chat.py [hunk 1]
```diff
@@ -201,13 +202,29 @@ def step(
...
+        # TODO: check for tooluse responses
 
         # log response and run tools
         if msg_response:
             yield msg_response.replace(quiet=True)
             yield from execute_msg(msg_response, confirm)
+            if toolsapi:
+                toolcall_re = re.compile(r"@(\w+): (\{.*?\})")
+                # if matches exactly
+                if match := toolcall_re.match(msg_response.content):
```
[the remark this answers is not in the document]
Comment (ErikBjare) [reply gptme/chat.py]: @ellipsis-dev That's as intended

**Comment by jrmi** [comment]
Hey @ErikBjare, are you planning to continue this soon? If not, would you be okay with me picking it up (on top of your changes obviously)? I believe this MR could significantly improve model performance, especially for smaller ones and prevent some random small bugs. :pray:

**Comment by ErikBjare** [comment]
@jrmi I'm kinda working on other stuff right now. No immediate plans to pick this up.

Feel free to give it a shot!

**Comment by ErikBjare** [comment]
Was merged in https://github.com/ErikBjare/gptme/pull/300, with much appreciated help from @jrmi!

**Pull request closed without merging**